apps/aecore/lib/aecore/contract/contract.ex

```diff
@@ -0,0 +1,167 @@
+defmodule Aecore.Contract.Contract do
+  @moduledoc """
+  Aecore contact module implementation.
+  """
+  alias Aecore.Contract.Contract
+  alias Aecore.Chain.Identifier
+  alias Aeutil.Hash
+
+  @version 1
+
+  # The @store_prefix is used to name the storage tree and keep
+  # all storage nodes in one subtree under the contract tree.
+  @store_prefix 16
+
+  @type contract :: %Contract{
+          id: Identifier.t(),
+          owner: Identifier.t(),
+          vm_version: byte(),
+          code: binary(),
+          store: %{binary() => binary()},
+          log: binary(),
+          active: boolean(),
+          referers: [Identifier.t()],
+          deposit: non_neg_integer()
+        }
+
+  @type t :: contract()
+
+  defstruct [:id, :owner, :vm_version, :code, :store, :log, :active, :referers, :deposit]
+
+  use Aecore.Util.Serializable
+
+  @spec new(Keys.pubkey(), non_neg_integer(), byte(), binary(), non_neg_integer()) :: contract()
+  def new(owner, nonce, vm_version, code, deposit) do
+    contract_id = create_contract_id(owner, nonce)
+    identified_contract = Identifier.create_identity(contract_id, :contract)
+    identified_owner = Identifier.create_identity(owner, :account)
+
+    %Contract{
+      id: identified_contract,
+      owner: identified_owner,
+      vm_version: vm_version,
+      code: code,
+      store: %{},
+      log: <<>>,
+      active: true,
+      referers: [],
+      deposit: deposit
+    }
+  end
+
+  @spec encode_to_list(Contract.t()) :: list()
+  def encode_to_list(%Contract{
+        owner: owner,
+        vm_version: vm_version,
+        code: code,
+        log: log,
+        active: active,
+        referers: referers,
+        deposit: deposit
+      }) do
+    encoded_active =
+      case active do
+        true -> 1
+        false -> 0
+      end
+
+    encoded_referers =
+      Enum.map(referers, fn referer ->
+        Identifier.encode_to_binary(referer)
+      end)
+
+    [
+      @version,
+      Identifier.encode_to_binary(owner),
+      :binary.encode_unsigned(vm_version),
+      code,
+      log,
+      encoded_active,
+      encoded_referers,
+      :binary.encode_unsigned(deposit)
+    ]
+  end
+
+  @spec decode_from_list(integer(), list()) :: {:ok, t()} | {:error, String.t()}
+  def decode_from_list(@version, [
+        owner,
+        vm_version,
+        code,
+        log,
+        active,
+        referers,
+        deposit
+      ]) do
+    decoded_active = decode_active(active)
+
+    decoded_referers =
+      Enum.reduce_while(referers, [], fn referer, acc ->
+        with {:ok, decoded_referer} <- Identifier.decode_from_binary(referer) do
+          [decoded_referer | acc]
+        else
+          _ ->
+            {:halt,
+             {:error,
+              "#{__MODULE__}: decode_from_list: Invalid contract referer: #{inspect(referer)}"}}
+        end
+      end)
+
+    case decoded_active do
```

Comment:
This here doesn't look really good. In this case you could use the `with` statement very effectively, but in order to do so the func `decode_active` should return: `{:ok, true}` and `{:ok, false}` so that you catch it in the with like so: `{:ok, _} <- decode_active`.

The decoded_referers you could catch like: `[_] <- decoded_referers`, because you will either end up with list of values or with a simple error variable (This depends on how you do the `reduce_while` func).

Doing this you will save much lines and make it more readable/understandable